### PR TITLE
feat: canal transit endpoints

### PR DESCRIPTION
**Pull request by jakubbuskiewicz**
#### RELATED TICKETS

https://vortexa.atlassian.net/browse/RND-6001

#### CHANGELOG

- added new endpoints for Canal Transit (search, time-series)
- updated docs
- added `Try me out` notebooks

#### TESTS

- proper tests were added

#### COMMENTS

- any other business

### CODE REVIEW

File: vortexasdk/endpoints/canal_transit_result.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    "direction",
+    "lock",
+    "queue_arrival_time",
+    "canal_entry_time",
+    "cargoes.0.product.label",
+    "cargoes.0.quantity_barrels",
+    "cargoes.0.quantity_tonnes",
+    "origin.0.label",
+    "destination.0.label",
+]
```
Comment (jakubbuskiewicz) [inline vortexasdk/endpoints/canal_transit_result.py:36]: Once we merge publish typo fix one more col will be added - `quantity_cubic_metres`

File: vortexasdk/endpoints/canal_transit.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
+from datetime import datetime, timedelta
```
Comment (jakubbuskiewicz) [inline vortexasdk/endpoints/canal_transit.py:1]: Binder links will be added once we merge a PR, I want to make sure we'll get proper URLs and Binder is reading from `master` branch to generate proper files and links

**Review by ministrudels — DISMISSED** [review]

File: .vscode/settings.json [hunk 3]
```diff
@@ -4,6 +4,13 @@
   "python.testing.unittestEnabled": false,
   "python.testing.pytestEnabled": true,
   "[python]": {
-    "editor.defaultFormatter": "ms-python.black-formatter"
-  }
+    "editor.defaultFormatter": "ms-python.autopep8"
```
[the remark this answers is not in the document]
Comment (jakubbuskiewicz) [reply .vscode/settings.json:7]: aww, I've missed that!

File: docs/endpoints/about-endpoints.md [hunk 4]
```diff
@@ -5,19 +5,21 @@ The endpoints module allows you to query Vortexa's data.
...
+1. Cargo Time Series
+1. EIA Forecasts
+1. Tonne-miles
+1. Vessel Availability
+1. Crude Onshore Inventories
+1. Freight Pricing
+1. Vessel Summary
+1. Vessel Positions
+1. Canal transit
+1. Canal transit Time Series
```
Comment (SeanBarry) [inline docs/endpoints/about-endpoints.md:22]: All numbers are 1.

Comment (jakubbuskiewicz) [reply docs/endpoints/about-endpoints.md:22]: Screenshot 2024-04-09 at 10 25 20

Markdown is able to loop through this. `1` is easier to track if you add/remove things as you do not need to update whole list

File: tests/endpoints/test_canal_transit_timeseries_real.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        monthlyBuckets = CanalTransitTimeseries().search(
+            time_min=datetime(2023, 9, 5),
+            time_max=datetime(2023, 12, 7),
+            metric="count_of_vessels",
+            timeseries_activity="started_waiting",
+            timeseries_frequency="week",
+        )
+
+        assert len(dailyBuckets) == 6
+        assert len(monthlyBuckets) == 14
```
Comment (SeanBarry) [inline tests/endpoints/test_canal_transit_timeseries_real.py:45]: I'm not familiar with this dataset, so just want to check - are these counts likely to change and break this test?

File: vortexasdk/endpoints/canal_transit.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+    A detailed explanation of what Canal Transit is can be found [here](https://docs.vortexa.com/reference/intro-canal-transit).
+    """
+
+    def __init__(self):
+        Search.__init__(self, CANAL_TRANSIT_SEARCH)
+
+    def search(
+        self,
+        filter_vessel_cubic_capacity_min: int = None,
```
Comment (SeanBarry) [inline vortexasdk/endpoints/canal_transit.py:27]: All of these types need to be set to Optional if they are being initialised as None

File: vortexasdk/endpoints/canal_transit_timeseries.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+logger = get_logger(__name__)
+
+
+class CanalTransitTimeseries(Search):
+    def __init__(self):
+        Search.__init__(self, CANAL_TRANSIT_TIME_SERIES)
+
+    def search(
+        self,
+        filter_vessel_cubic_capacity_min: int = None,
```
Comment (SeanBarry) [inline vortexasdk/endpoints/canal_transit_timeseries.py:22]: All of these types need to be set to Optional if they are being initialised as None

File: vortexasdk/endpoints/canal_transit.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            ),
+            "filter_charterers": convert_to_list(exclude_filter_charterer),
+            "filter_effective_controllers": convert_to_list(
+                exclude_filter_effective_controller
+            ),
+        }
+
+        api_params: Dict[str, Any] = {
+            "filter_booked_time_max": to_ISODate(filter_booked_time_max)
+            if filter_booked_time_max
```
Comment (SeanBarry) [inline vortexasdk/endpoints/canal_transit.py:151]: This formatting is a bit weird, it took me a few secs looking at it to work out what was going on. Has this been passed through Black?

File: docs/endpoints/about-endpoints.md (hunk 4 above)
Comment (SeanBarry) [reply docs/endpoints/about-endpoints.md:22]: Today I learned!

File: tests/endpoints/test_canal_transit_timeseries_real.py (hunk 5 above)
Comment (jakubbuskiewicz) [reply tests/endpoints/test_canal_transit_timeseries_real.py:45]: in this case no - we're testing buckets returned by api and it's calculated based on number of days and weeks provided by by min/max dates

File: vortexasdk/endpoints/canal_transit.py (hunk 8 above)
Comment (jakubbuskiewicz) [reply vortexasdk/endpoints/canal_transit.py:151]: yeah, I know, had the same feeling! Unfortunately that's how Black formatted it

File: tests/endpoints/test_canal_transit_timeseries_real.py (hunk 5 above)
Comment (SeanBarry) [reply tests/endpoints/test_canal_transit_timeseries_real.py:45]: Makes sense, thanks for clarifying!

File: vortexasdk/endpoints/canal_transit.py (hunk 8 above)
Comment (SeanBarry) [reply vortexasdk/endpoints/canal_transit.py:153]: Roger that!

**Review by SeanBarry — APPROVED** [review]
LGTM!

**Pull request merged**